### ISSUE TITLE
new strategy like "free" but uninterrupted execution per host

### ISSUE THREAD

**Issue by backbord** [ansible/ansible#39374]
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

plugins/strategy/atomic

##### ANSIBLE VERSION

```
ansible 2.5.0
  config file = /home/tim/test/ansible/ansible.cfg
  configured module search path = ['/home/tim/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/tim/conda/envs/ansible/lib/python3.6/site-packages/ansible
  executable location = /home/tim/conda/envs/ansible/bin/ansible
  python version = 3.6.5 | [...] | (default, Apr 13 2018, 12:47:22) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```

##### CONFIGURATION

```
ANSIBLE_NOCOWS(/home/tim/test/ansible/ansible.cfg) = True
```
(Output of ``ansible-config dump --only-changed``)

##### OS / ENVIRONMENT

N/A

##### SUMMARY

I would like to have a strategy that executes playbooks in such a way that once the play for a host is started, it will run to completion without "its" worker being snatched for plays of other hosts.
In the following, I will call the desired strategy "atomic".

This is closely related to the behaviour of the "free" strategy, so let's have a look at the execution order of tasks when run with the free strategy.

As a guiding example, consider a playbook with three tasks, "stop" followed by "update" (which takes some time), followed by "start", run on an inventory with three hosts "A", "B" and "C" with ``--forks 1``.

In this setup, I would like the "atomic" strategy to have all tasks for host A will be executed before any task for host B (there is only 1 fork!), so the order will be
1. A stop
2. A update
3. A start
4. B stop
5. B update
6. B start
7. C stop
8. C update
9. C start

In contrast to the behaviour described above, the "free" strategy will execute in this order:

1. A stop
2. B stop
3. C stop
4. A update
5. B update
6. A start
7. C update
8. B start
9. C start

Here, A's play is interrupted by tasks for B and C.
There is no waiting of host B for a new slot that would be opened once host A is done with the play (and C waiting for B's slot).

With two workers ``--forks 2``, I expect an execution order for "atomic" like the following (parallel execution as unordered list):
1. - A stop
   - B stop
2. - A update
   - B update
3. - A start
   - B start
4. - C stop
5. - C update
6. - C start

Whereas the "free" strategy will execute in an order that is more like the following:

1. - A stop
   - B stop
2. - C stop
   - A update
3. - B update
   - A start
4. - B start
   - C update
5. - C start

We observe that "C stop" is executed much earlier than expected, so that from C's perspective the play isn't executed as fast as possible and it did not wait for a slot (from A or B) to open up.

##### STEPS TO REPRODUCE

```yaml
- hosts: localhost
  gather_facts: no
  tasks:
    - add_host: name="{{ item }}"
      no_log: yes
      loop: [A, B, C]

- hosts: all
  gather_facts: no
  strategy: free
  tasks:
    - name: stop
      local_action:
        module: shell
        cmd: "date -u +\"stop {{ inventory_hostname }} %H:%M:%S.%N\""
      register: msg

    - debug: msg="{{ msg.stdout }}"
    - name: update
      local_action:
        module: shell
        cmd: "date -u +\"updating {{ inventory_hostname }} %H:%M:%S.%N\" ; sleep 2 ; date -u +\"updated {{ inventory_hostname }} %H:%M:%S.%N\""
      register: msg

    - debug: msg="{{ msg.stdout }}"

    - name: start
      local_action:
        module: shell
        cmd: "date -u +\"start {{ inventory_hostname }} %H:%M:%S.%N\""
      register: msg

    - debug: msg="{{ msg.stdout }}"
```

Output of play with ``--forks 1``
```
$ ansible-playbook test.yml --forks 1
 [WARNING]: provided hosts list is empty, only localhost is available. Note
that the implicit localhost does not match 'all'


PLAY [localhost] ***************************************************************

TASK [add_host] ****************************************************************
changed: [localhost] => (item=None)
changed: [localhost] => (item=None)
changed: [localhost] => (item=None)

PLAY [all] *********************************************************************

TASK [stop] ********************************************************************
changed: [A -> localhost]
changed: [B -> localhost]
changed: [C -> localhost]

TASK [debug] *******************************************************************
ok: [A] => {
    "msg": "stop A 07:02:44.174029962"
}
ok: [B] => {
    "msg": "stop B 07:02:44.436160558"
}

TASK [update] ******************************************************************
changed: [A -> localhost]

TASK [debug] *******************************************************************
ok: [C] => {
    "msg": "stop C 07:02:44.690951127"
}

TASK [debug] *******************************************************************
ok: [A] => {
    "msg": "updating A 07:02:45.028612055\nupdated A 07:02:47.031238359"
}

TASK [update] ******************************************************************
changed: [B -> localhost]

TASK [start] *******************************************************************
changed: [A -> localhost]

TASK [update] ******************************************************************
changed: [C -> localhost]

TASK [debug] *******************************************************************
ok: [A] => {
    "msg": "start A 07:02:49.625430719"
}

TASK [debug] *******************************************************************
ok: [B] => {
    "msg": "updating B 07:02:47.366121959\nupdated B 07:02:49.367728418"
}
ok: [C] => {
    "msg": "updating C 07:02:49.882540427\nupdated C 07:02:51.885279434"
}

TASK [start] *******************************************************************
changed: [B -> localhost]
changed: [C -> localhost]

TASK [debug] *******************************************************************
ok: [B] => {
    "msg": "start B 07:02:52.251900634"
}
ok: [C] => {
    "msg": "start C 07:02:52.496086386"
}

PLAY RECAP *********************************************************************
A                          : ok=6    changed=3    unreachable=0    failed=0
B                          : ok=6    changed=3    unreachable=0    failed=0
C                          : ok=6    changed=3    unreachable=0    failed=0
localhost                  : ok=1    changed=1    unreachable=0    failed=0
```

Output of play with ``--forks 2``:
```
$ ansible-playbook test.yml --forks 2
 [WARNING]: provided hosts list is empty, only localhost is available. Note
that the implicit localhost does not match 'all'


PLAY [localhost] ***************************************************************

TASK [add_host] ****************************************************************
changed: [localhost] => (item=None)
changed: [localhost] => (item=None)
changed: [localhost] => (item=None)

PLAY [all] *********************************************************************

TASK [stop] ********************************************************************
changed: [B -> localhost]
changed: [A -> localhost]

TASK [debug] *******************************************************************
ok: [A] => {
    "msg": "stop A 07:16:45.489372102"
}
ok: [B] => {
    "msg": "stop B 07:16:45.489355132"
}

TASK [stop] ********************************************************************
changed: [C -> localhost]

TASK [update] ******************************************************************
changed: [A -> localhost]

TASK [debug] *******************************************************************
ok: [C] => {
    "msg": "stop C 07:16:45.736580915"
}

TASK [debug] *******************************************************************
ok: [A] => {
    "msg": "updating A 07:16:45.811320486\nupdated A 07:16:47.813289791"
}

TASK [update] ******************************************************************
changed: [B -> localhost]

TASK [start] *******************************************************************
changed: [A -> localhost]

TASK [debug] *******************************************************************
ok: [B] => {
    "msg": "updating B 07:16:45.969581878\nupdated B 07:16:47.971922983"
}

TASK [debug] *******************************************************************
ok: [A] => {
    "msg": "start A 07:16:48.204427216"
}

TASK [start] *******************************************************************
changed: [B -> localhost]

TASK [debug] *******************************************************************
ok: [B] => {
    "msg": "start B 07:16:48.523619978"
}

TASK [update] ******************************************************************
changed: [C -> localhost]

TASK [debug] *******************************************************************
ok: [C] => {
    "msg": "updating C 07:16:48.132759023\nupdated C 07:16:50.135063889"
}

TASK [start] *******************************************************************
changed: [C -> localhost]

TASK [debug] *******************************************************************
ok: [C] => {
    "msg": "start C 07:16:50.444562398"
}

PLAY RECAP *********************************************************************
A                          : ok=6    changed=3    unreachable=0    failed=0
B                          : ok=6    changed=3    unreachable=0    failed=0
C                          : ok=6    changed=3    unreachable=0    failed=0
localhost                  : ok=1    changed=1    unreachable=0    failed=0
```

---

Originally, I considered this a documentation bug or behavioural bug, but @jimi-c proposed in #39376 that this would be better suited as a novel strategy, so I'll try my best at a new PR. :-)

I'll see to commit a new PR.

Thanks,
Tim

PS: I changed this description from a bug report to a feature idea. I hope that's ok.

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/plugins/strategy/free.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/free.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/plugins/strategy/__init__.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/__init__.py)
* [lib/ansible/plugins/strategy/debug.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/debug.py)
* [lib/ansible/plugins/strategy/free.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/free.py)
* [lib/ansible/plugins/strategy/linear.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/linear.py)
* [test/units/plugins/strategy/__init__.py](https://github.com/ansible/ansible/blob/devel/test/units/plugins/strategy/__init__.py)
* [test/units/plugins/strategy/test_strategy_base.py](https://github.com/ansible/ansible/blob/devel/test/units/plugins/strategy/test_strategy_base.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/plugins/strategy/__init__.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/__init__.py)
* [lib/ansible/plugins/strategy/debug.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/debug.py)
* [lib/ansible/plugins/strategy/free.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/free.py)
* [lib/ansible/plugins/strategy/linear.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/linear.py)
* [test/units/plugins/strategy/__init__.py](https://github.com/ansible/ansible/blob/devel/test/units/plugins/strategy/__init__.py)
* [test/units/plugins/strategy/test_strategy_base.py](https://github.com/ansible/ansible/blob/devel/test/units/plugins/strategy/test_strategy_base.py)
* [test/units/plugins/strategy/test_strategy_linear.py](https://github.com/ansible/ansible/blob/devel/test/units/plugins/strategy/test_strategy_linear.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by backbord** [comment]
Closed via #44586
